support/agent/cases_test.go

```diff
@@ -177,7 +185,150 @@ func TestCommitPowerAndCheckInvariants(t *testing.T) {
+		Seed:             rnd.Int63(),
+		CheckpointEpochs: 1000,
+	}, agent.CreateMinerParamsV2)
+
+	// create miners
+	workerAccounts := vm_test2.CreateAccounts(ctx, t, v, minerCount, initialBalance, rnd.Int63())
+	sim.AddAgent(agent.NewMinerGenerator(
+		workerAccounts,
+		agent.MinerAgentConfig{
+			PrecommitRate:    2.0,
+			FaultRate:        0.00001,
+			RecoveryRate:     0.0001,
+			UpgradeSectors:   true,
+			ProofType:        abi.RegisteredSealProof_StackedDrg32GiBV1_1,
+			StartingBalance:  big.Div(initialBalance, big.NewInt(2)),
+			MinMarketBalance: big.NewInt(1e18),
+			MaxMarketBalance: big.NewInt(2e18),
+		},
+		1.0, // create miner probability of 1 means a new miner is created every tick
+		rnd.Int63(),
+	))
+
+	clientAccounts := vm_test2.CreateAccounts(ctx, t, v, clientCount, initialBalance, rnd.Int63())
+	agent.AddDealClientsForAccounts(sim, clientAccounts, rnd.Int63(), agent.DealClientConfig{
+		DealRate:         .05,
+		MinPieceSize:     1 << 29,
+		MaxPieceSize:     32 << 30,
+		MinStoragePrice:  big.Zero(),
+		MaxStoragePrice:  abi.NewTokenAmount(200_000_000),
+		MinMarketBalance: big.NewInt(1e18),
+		MaxMarketBalance: big.NewInt(2e18),
+	})
+
+	// Run v2 for 5000 epochs
+	var pwrSt power2.State
+	for i := 0; i < 5000; i++ {
+		require.NoError(t, sim.Tick())
+		epoch := sim.GetVM().GetEpoch()
+		if epoch%100 == 0 {
+			stateTree, err := states2.LoadTree(sim.GetVM().Store(), sim.GetVM().StateRoot())
+			require.NoError(t, err)
+
+			totalBalance, err := sim.GetVM().GetTotalActorBalance()
+			require.NoError(t, err)
+
+			acc, err := states2.CheckStateInvariants(stateTree, totalBalance, sim.GetVM().GetEpoch()-1)
+			require.NoError(t, err)
+			require.True(t, acc.IsEmpty(), strings.Join(acc.Messages(), "\n"))
+
+			require.NoError(t, sim.GetVM().GetState(builtin.StoragePowerActorAddr, &pwrSt))
+
+			// assume each sector is 32Gb
+			sectorCount := big.Div(pwrSt.TotalBytesCommitted, big.NewInt(32<<30))
+
+			fmt.Printf("Power at %d: raw: %v  cmtRaw: %v  cmtSecs: %d  cnsMnrs: %d avgWins: %.3f  msgs: %d\n",
+				epoch, pwrSt.TotalRawBytePower, pwrSt.TotalBytesCommitted, sectorCount.Uint64(),
+				pwrSt.MinerAboveMinPowerCount, float64(sim.WinCount)/float64(epoch), sim.MessageCount)
+		}
+	}
+
+	// Migrate
+	v2 := sim.GetVM()
+	log := nv10.TestLogger{TB: t}
+	priorEpoch := v2.GetEpoch() - 1 // on tick sim internally creates new vm with epoch set to the next one
+	nextRoot, err := nv10.MigrateStateTree(ctx, v2.Store(), v2.StateRoot(), priorEpoch, nv10.Config{MaxWorkers: 1}, log, nv10.NewMemMigrationCache())
+	require.NoError(t, err)
+
+	lookup := map[cid.Cid]rt.VMActor{}
+	for _, ba := range exported.BuiltinActors() {
+		lookup[ba.Code()] = ba
+	}
+
+	v3, err := vm_test.NewVMAtEpoch(ctx, lookup, v2.Store(), nextRoot, priorEpoch+1)
+	require.NoError(t, err)
+
+	stateTree, err := v3.GetStateTree()
+	require.NoError(t, err)
+	totalBalance, err := v3.GetTotalActorBalance()
```

Comment:
Shouldn't this be the balance from _before_ the migration?

Reply:
No I think this should be balance after.  This call is checking that the v3 state is internally consistent post migration, it's not asserting the correctness of the migration transition.

We could have used v2 balance to do this because the migration function should not have changed total balance.

Reply:
Then I don't think this is useful. GetTotalActorBalance just does the exact same thing as CheckStateInvariants.

Reply:
I don't follow.  We need the output of `GetTotalActorBalance` as input to `CheckStateInvariants`.  Are you saying this    should be called inside CheckStateInvariants instead of outside?  If so I agree but it belongs in its own PR.

Reply:
To clarify: "This call is checking that the v3 state is internally consistent post migration, it's not asserting the correctness of the migration transition." refers to CheckStateInvariants, not GetTotalActorBalance(), my bad on the confusing wording.